React autosave: patient dashboard form (bugs fixed 🙂 )

mercedesb opened the pull request:
I rule and have completed some work on Case Manager that's ready for review!

This PR rewrites the patient dashboard form in React with safe autosave. This is mostly the same as #3216 but with the identified bugs fixed.

Let me know if anything else pops up. I'm sorry for the long delay on this 🫣 

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.

lomky commented:
I will get this into sandbox and tested shortly!

Fixes #3240

colinxfleming commented on test/system/update_patient_info_test.rb:
```diff
@@ -28,7 +28,6 @@ class UpdatePatientInfoTest < ApplicationSystemTestCase
     describe 'updating name' do
       before do
         fill_in 'First and last name', with: 'Susie Everyteen 2'
-        click_away_from_field
```
honestly dope

colinxfleming commented on app/javascript/src/components/PatientDashboardForm.jsx:
```diff
@@ -44,9 +43,9 @@ export default PatientDashboardForm = ({
-    return debounce(autosave, 300)
-  }, []);
+  const debouncedAutosave = (params) => {
+    return debounce(autosave(params), 300)
```
this works like a charm for me fysa

colinxfleming commented on app/views/patients/_patient_dashboard.html.erb:
```diff
@@ -1,77 +1,12 @@
+    weeksOptions: weeks_options.map { |opt| { label: opt[0], value: opt[1] } },
+    daysOptions: days_options.map { |opt| { label: opt[0], value: opt[1] } },
+    initialCallDate: patient.initial_call_date.strftime("%m/%d/%Y"),
+    statusHelpText: status_help_text(patient),
```
at some point we'll need to find a way to re-rack this from changes (easiest example on main: setting an appointment date makes the status `Fundraising`, removing appointment dates makes the status `Needs Appointment`. That said I'm perfectly fine letting this regress a bit and merging this, and tackling this later, because I'd like to start running this react code in prod again. (Please make an issue if we do that so we don't forget.)

mercedesb replied:
I'm not totally following your comment but I think that's b/c I'm not as familiar with all the parts of the app as you are. Can you clarify by what you mean by "this"? The status help text? Or the status input itself?

colinxfleming replied:
yeah, the status help text -- the content changes based on the state of the patient (and I believe that's a model field right now). On prod now it refreshes after a change without a page reload.

mercedesb replied:
cool cool, I went ahead and fixed that in this branch 🤗 hopefully no regressions now

colinxfleming commented on app/javascript/src/components/PatientDashboardForm.jsx:
```diff
@@ -26,7 +26,6 @@ export default PatientDashboardForm = ({
 
   const autosave = async (updatedData) => {
     const updatedPatientData = { ...patientData, ...updatedData }
-    setPatientData(updatedPatientData)
```
this is the main culprit right?

mercedesb replied:
actually the change below with the debounce was the main culprit. a combo of the useMemo and not passing the params through to the autosave function

colinxfleming replied:
ah, good to know! makes sense

mercedesb replied:
I really need to learn to just stop reaching for useMemo. I mess it up nearly every time and it has cost me hours of my life 😅

colinxfleming reviewed (commented):
this is looking pretty good to me. I want to have @lomky take a look at it real quick but then I'm comfortable rolling with it. One minor regression noted in a comment but (as noted) I'm good to roll with this.

Sorry again for the review delay!

mercedesb commented:
Looks like I'm running into test failures due to the debounced save and ajax stuff with Capybara. I'll continue looking into this this week.

lomky commented:
I'm seeing a small bug when I try to set "0 days" in the weeks along settings, it blanks it out. I suspect its something to do with the javascript -> ruby translation with "whats a false", but I haven't dug in more

mercedesb commented:
> I'm seeing a small bug when I try to set "0 days" in the weeks along settings, it blanks it out. I suspect its something to do with the javascript -> ruby translation with "whats a false", but I haven't dug in more

Great, I'll add that to my list when I'm resolving the specs issue

mercedesb commented:
> I'm seeing a small bug when I try to set "0 days" in the weeks along settings, it blanks it out

Looks like this is what currently happens on main and is not a regression introduced by this branch

mercedesb commented:
😮‍💨 ok I think we're good! finally. sorry for the avalanche of notifications. I've addressed all feedback
- dynamic status text regression
- blank '0 days' regression
- spec failures
- a host of mini bugs that we're not gonna talk about

have a great week!

lomky commented:
All the functionality is looking great! I think we have one last visual to fix where the Days Along field is shifting up, but otherwise it seems to be working great for me and i'm so excited to get it merged!

in the old component, it had an invisible label pushing it down to be level with Weeks Along.

main:
![image](https://github.com/user-attachments/assets/f9d459c5-4719-4e1d-be8e-282198d9fd73)

branch:
![image](https://github.com/user-attachments/assets/92cc2475-b5ed-4969-90e4-fef18d4f330f)

mercedesb commented:
Visual regression fixed.

I've rebased this branch on latest main but am getting a security vulnerability that needs to be fixed (the `json` gem) as well as specs related to auth which shouldn't be from anything this branch touches.

I'm going to let this be and hope that those issues are resolved in the base branch. Let me know if you need me to rebase or retrigger workflows this week.

colinxfleming commented:
Popping in here to say re: build, there's currently SOMETHING up with [github actions and capybara](https://github.com/teamcapybara/capybara/issues/2800), this results in about 10 system test failures and we can ignore those for now. I also bumped the json gem on main, so if the regression is solved for I think we can smash merge even if build's not passing.

mercedesb commented:
@lomky 

Re: "Field Changes sometimes 'saved' but then lost", I currently have a 300 ms debounce on the input. I'm not able to reproduce what you've described. But I can sometimes force a page reload faster than the 300 ms if I'm realllllllyyyyy trying. Could that be something you're running into? Seeing the green banner from a previous save and reloading the page before the new change can be committed? Want me to decrease the 300 ms debounce on the autosave?

Re: Days Along off-by-one
Can you clarify your screenshot? I'm trying to guess what it's supposed to say based on the timestamp of when you left your comment but I still don't know which part of the screenshot is correct and which isn't. Could you let me know what you expect to see in which box and where it's wrong? That would save me a lot of time tracking down whatever could be happening.

mercedesb commented on app/javascript/src/components/Input.jsx:
```diff
@@ -15,7 +15,15 @@ export default Input = ({
+    value = initialValue
+  } else {
+    ([value, setValue] = useState(initialValue || ""))
+  }
```
This is a good point and something I went back and forth on a lot. I don't want to mutate the state variable directly either since that also breaks React conventions.

I've pushed up a more complex but more complete solution in today's commits.

lomky commented:
Re: "Field Changes" - I'm not trying especially hard, just typing and tabbing through fields! but that is pretty fast.

Re: Days Along off-by-one - The value selected is 2, but the display (see the weeks along below) says 3. That's true of the other places the value showed, too. But when I checked the console, it was saved properly as 2. If I changed the value, the display _also_ changed, but still off by one. it was a weird bug and I couldn't reproduce it.

For overall context, I spent about 5-10 minutes just naviating around, typing, and changing things, and saw the off-by-one once and field change loss two or three times. So it wasn't even the most common occurence, but too often for my comfort, if that makes sense?

mercedesb commented:
I _think_ bugs are fixed? 

I'm getting `Selenium::WebDriver::Error::SessionNotCreatedError` errors in the specs which seem to be affecting all the recent PRs so those failures aren't introduced by these changes.

mercedesb commented:
Going to close this stale PR
